core/src/main/java/org/apache/accumulo/core/fate/Fate.java

```diff
@@ -425,11 +474,36 @@ public Fate(T environment, FateStore<T> store, boolean runDeadResCleaner,
           }
         }
         idleCountHistory.clear();
+      } else if (needed < 0) {
+        // If we need the pool to shrink, then ensure excess TransactionRunners are safely stopped.
```

Comment:
Maybe instead of trying to stop and then unstop thread and sleep, could instead just try to get stop set on a goal number of threads.  Do not wait for them to exit. Maybe something like the following.


```java
        int numStopped = runningTxRunners.stream().filter(tr->tr.isStopped()).count();
        int numToStop = -1*(numStopped + needed);
        for(var tr : runningTxRunners) {
          if(numToStop <= 0) {
            break;
          }
          
          // returns true if it was not already stopped
          if(tr.setStop()) {
            numToStop--;
          }
        }
```

Reply:
This is a great solution. Implemented in newest commit. Was originally worried about any one TransactionRunner taking too long to stop (the transaction it's working on takes a while) when there may be other TransactionRunners that can more easily be stopped. However, we can use the delay between executions of the fate pool watcher as our "wait" which I didn't consider. Eventually the correct number of TransactionRunners will be stopped which seems sufficient.